ext/drreg/drreg.c

```diff
@@ -498,10 +498,20 @@ drreg_insert_restore_all(void *drcontext, instrlist_t *bb, instr_t *inst,
         }
     }
     /* Before each app read, or at end of bb, restore spilled registers to app values.
-     * XXX i#5121: multi-phase use causes correctness issues here!  Our restore
-     * in the insertion phase here could actually restore the tool value from the
-     * app2app phase.  A better solution is needed for multi-phase, or perhaps
-     * disallowing the use of the same register.
+     * XXX i#5121: For regs that are in spilled state at this instr also during the
+     * app2app phase, the following does not restore the app value, it restores the
+     * app2app meta value instead. This is expected and correct, because the drreg use
+     * in insertion phase adds spills/restores around the drreg spills/restores from
+     * previous phases like they were app instrs. So, the app2app meta value is no
+     * different than the app value for the insertion phase.
```

Comment:
I found this confusing: it would only be the app2app meta value if the app2app phase does *not* restore before the read.  The last sentence only applies if the app2app phase restored before the read.

Reply:
>  it would only be the app2app meta value if the app2app phase does not restore before the read

This comment starts with "For regs that are in spilled state at this instr also during the app2app phase", so it's talking about the not-restored case.

> The last sentence only applies if the app2app phase restored before the read.

I reworded it. Is it better now?

Reply:
Resolved in newer comment thread.